homeassistant/components/supla/__init__.py

```diff
@@ -0,0 +1,158 @@
+
+    for server_name, server in hass.data[SUPLA_SERVERS].items():
+
+        for channel in server.get_channels(include=['iodevice']):
+            channel_function = channel['function']['name']
+            component_name = SUPLA_FUNCTION_HA_CMP_MAP.get(channel_function)
+
+            if component_name is None:
+                _LOGGER.warning(
+                    'Unsupported function: %s, channel id: %s',
+                    channel_function, channel['id']
+                )
+                continue
+
+            channel['server_name'] = server_name
+            component_configs.setdefault(component_name, []).append(channel)
+
+    # Load discovered devices
+    for component_name, channel in component_configs.items():
+        load_platform(
+            hass,
+            component_name,
+            'supla',
+            channel,
+            hass_config
+        )
+
+
+def setup(hass, base_config):
+    """Set up the Supla component."""
+    from pysupla import SuplaAPI
+
+    server_confs = base_config[DOMAIN][CONF_SERVERS]
+
+    hass.data[SUPLA_SERVERS] = {}
+    hass.data[SUPLA_CHANNELS] = {}
+
+    for server_conf in server_confs:
+        server = SuplaAPI(
+            server_conf[CONF_SERVER],
+            server_conf[CONF_ACCESS_TOKEN]
+        )
+        hass.data[SUPLA_SERVERS][server_conf[CONF_SERVER]] = server
+
+    discover_devices(hass, base_config)
+
+    return True
+
+
+class SuplaChannel(Entity):
+    """Base class of a Supla Channel (an equivalent of HA's Entity)."""
+
+    def __init__(self, channel_data):
+        """Channel data -- raw channel infor from PySupla."""
+        self.server_name = channel_data['server_name']
+        self.channel_data = channel_data
+
+    @property
+    def server(self):
+        """Return PySupla's server component associated with entity."""
+        return self.hass.data[SUPLA_SERVERS][self.server_name]
+
+    @property
+    def hidden(self) -> bool:
+        """Return True if the entity should be hidden from UIs."""
+        return self.channel_data['hidden']
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return 'supla_{}_{}'.format(self.server_name, self.channel_data['id'])
+
+    @property
+    def name(self) -> Optional[str]:
+        """Return the name of the device."""
+        return self.channel_data['caption']
+
+    @property
+    def should_poll(self):
+        """Supla's web API requires polling."""
+        return True
+
+    def action(self, action, **add_pars):
+        """
+        Run server action.
+
+        Actions are currently hardoced in components.
+        Supla's API enables autodiscovery
+        """
+        _LOGGER.debug(
+            'Executing action %s on channel %d, params: %s',
+            action,
+            self.channel_data['id'],
+            add_pars
+        )
+        self.server.execute_action(self.channel_data['id'], action, **add_pars)
+        self.update()
```

Comment:
Update will be called at the end of the service call for polling entities. We can remove this call here.

Reply:
Fixed.